src/main/java/org/osc/manager/rest/server/api/SecurityGroupInterfaceApis.java

```diff
@@ -0,0 +1,150 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.osc.manager.rest.server.api;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.persistence.EntityManager;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.osc.manager.ism.api.IsmSecurityGroupInterfaceApi;
+import org.osc.manager.ism.api.util.ValidationUtil;
+import org.osc.manager.ism.entities.DeviceEntity;
+import org.osc.manager.ism.entities.SecurityGroupEntity;
+import org.osc.manager.ism.entities.SecurityGroupInterfaceEntity;
+import org.osc.manager.ism.model.SecurityGroupInterfaceElementImpl;
+import org.osc.manager.ism.model.VirtualSystemElementImpl;
+import org.osc.sdk.manager.element.ManagerSecurityGroupInterfaceElement;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.transaction.control.TransactionControl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Component(service = SecurityGroupInterfaceApis.class)
+@Path("/devices/{deviceId}/securityGroupInterfaces")
+@Consumes({ MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML })
+@Produces({ MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML })
+public class SecurityGroupInterfaceApis {
+
+    private static Logger LOG = LoggerFactory.getLogger(SecurityGroupInterfaceApis.class);
+
+    private IsmSecurityGroupInterfaceApi sgiApi;
+
+    private TransactionControl txControl;
+
+    private EntityManager em;
+
+    private ValidationUtil validationUtil;
+
+    public void init(EntityManager em, TransactionControl txControl) throws Exception {
+        this.txControl = txControl;
+        this.em = em;
+        this.validationUtil = new ValidationUtil(this.txControl, this.em);
+    }
+
+    @POST
+    public String createSecurityGroupInterface(@PathParam("deviceId") Long deviceId,
+            SecurityGroupInterfaceEntity entity)
+                    throws Exception {
+        LOG.info(String.format("Creating the security group interface with name %s ", entity.getName()));
+
+        DeviceEntity device = this.validationUtil.getDeviceOrThrow(Long.toString(deviceId));
+        this.validationUtil.validateParentIdMatches(device, Long.parseLong(entity.getDevice().getId()),
+                "SecurityGroupInterface");
+
+        VirtualSystemElementImpl vs = new VirtualSystemElementImpl(deviceId, null);
+        this.sgiApi = new IsmSecurityGroupInterfaceApi(vs, null, this.txControl, this.em);
+
+        if (entity.getSecurityGroup() != null) {
```

Comment:
generally the pattern is that the security group is created before the SGI...

or atleast the security group creation/updation should be independent of SGI creation. We should validate the sg exists and then add reference to it as applicable. "new SecurityGroupEntity()" should not be needed